app/models/runs_reducers.rb

```diff
@@ -60,9 +56,25 @@ def reduce(subject_id, user_id, extract_ids=[], and_check_rules: false)
     raise ReductionConflict, "Transient uniqueness violation"
   end
 
+  def prepare_extract_query(subject_id, user_id)
+    { subject_id: subject_id, user_id: user_id }.tap do |extract_query|
+      case reducible
+      when Workflow
+        extract_query[:workflow_id] = reducible.id
+      when Project
+        extract_query[:project_id] = reducible.id
+      end
+    end
+  end
+
+  def prepare_reduction_query(subject_id, user_id)
+    { reducible_id: reducible.id, reducible_type: reducible.class.to_s, subject_id: subject_id, user_id: user_id }
+  end
+
   def persist_reductions(reductions)
     ActiveRecord::Base.transaction do
       reductions.each do |reduction|
+        # never save user reductions with no user id because that doesn't make sense
```

Comment:
not blocking: could this be a validation on the reduction record? Also it seems like a waste to compute the data for this reduction only to throw it away and then probably do it again at some point....is there a way to avoid this reduction cost instead of just avoiding saving it?

Reply:
There is a validation on the reduction record in the database; I added this line to make it a no-op instead of logging an error to Sentry/Rollbar when it happens. We do attempt at several other points to avoid reducing in this situation; this line represents part of a defense in depth against the behavior.